### PR TITLE
Jiajun/build system refactor

### PR DESCRIPTION
**Pull request by Jiajun-Huang**
### Summary
Small change on some CMakelist.txt, allowing the different boards to include or exclude the code in shared directory.
It solves the dependency issue in shared code when one shared code depends on a specific driver that some boards don't have., only the board has the driver gets compiled and others boards will not.
Excluding unnecessary code will also reduce the compiled binary size.

### Changelist 
Add two parameter in `create_arm_binary` and `create_arm_binary_or_tests_for_board` functions in `firmware/CMakeList.txt`
Change the CMakeList.txt for each board accordingly.

### Testing Done
Remove cache in build_arm and build_x64 
Load Cmake test and embedded
Build all embedded binary, all test and shared test

### Checklist
- [x ] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [ ] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

**Comment by Jiajun-Huang** [comment]
The issue in ‘scripts/utilities/supported_boards.py’ seems like just a Python version issue. I can remove the fix later.

File: firmware/thruna/BMS/CMakeLists.txt
Comment (Lucien950) [inline firmware/thruna/BMS/CMakeLists.txt:63]: can we set this in `firmware/CMakeLists.txt` rather than pass it all the way down the function? This saves a lot more code between all the files in `firemware/thruna/<BOARD>/CmakeLists.txt`

Comment (Lucien950) [reply firmware/thruna/BMS/CMakeLists.txt:63]: Because from what I can tell the variables are the same throughout.

Comment (Lucien950) [reply firmware/thruna/BMS/CMakeLists.txt:63]: By the way variables defined at higher levels of the folder structure carry over to lower levels.

**Review by Lucien950 — CHANGES_REQUESTED** [review]

File: firmware/thruna/BMS/CMakeLists.txt
Comment (Jiajun-Huang) [reply firmware/thruna/BMS/CMakeLists.txt:63]: The primary I made this change is this.
I am currently writing an SD_card driver that is based on the HAL SD_Card driver for f4dev. Gus told me to put this in the shared folder and I think it makes sense because other boards might use this code in the future.  But, if I set it up like this, all other boards can't compile because they do not have the HAL driver for the SD card. 

Also, I am setting up a new board f4dev and I only want it to have HAL SD_Card driver on the project for now. But, if I want to build the binary, I have to add a bunch of other drivers and middleware to solve the dependency issue just to compile the shared code.  

So I want each individual board to have the ability to select the shared code it needs; so that other existing boards can include everything in shared code but exclude the SD_Card driver, and the new f4dev board can only include the SD_card driver and exclude anything else.

In order to do that, I think it has to be done in the CmakeLists.txt for each board.

Comment (Lucien950) [reply firmware/thruna/BMS/CMakeLists.txt:63]: Can you make it so that `SHARED_APP_SRCS`, `SHARED_IO_SRCS`, `SHARED_HW_SRCS` in `firmare/CMakeLists.txt` and set `ALL_SHARED_SRCS` and `ALL_SHARED_INCLUDE_DIRS` in each CMakeLists.txt?

Comment (Lucien950) [reply firmware/thruna/BMS/CMakeLists.txt:63]: So you pick everything you want from each board's CMakeLists.txt but you keep the helper variables in the higher context.

Comment (Jiajun-Huang) [reply firmware/thruna/BMS/CMakeLists.txt]: The changes are pushed.

**Review by Lucien950 — APPROVED** [review]
The code looks good. Please wait until @gtaharaedmonds reviews to merge to master.

File: firmware/CMakeLists.txt
Comment (gtaharaedmonds) [inline firmware/CMakeLists.txt:149]: These should probably be renamed to `SELECTED_ARM_BINARY_*`

File: firmware/thruna/BMS/CMakeLists.txt
Comment (gtaharaedmonds) [inline firmware/thruna/BMS/CMakeLists.txt:70]: nit: Can you make whitespace consistent on the board files? i.e. remove the double newlines

File: firmware/thruna/BMS/CMakeLists.txt
Comment (gtaharaedmonds) [inline firmware/thruna/BMS/CMakeLists.txt:68]: Is this unnecessary - is there ever a case where a certain board will want to ignore  a certain include directory? 
I think every board will want to look in `app`, `io`, and `hw`. I believe headers only get included in the binary if a C file `#include`s it so this shouldn't affect the binary size either. Can you think if we will ever need this?

**Review by gtaharaedmonds — COMMENTED** [review]
Looks great. We should probably do this for tests too but that can be a follow up since the SD driver is higher priority

File: firmware/thruna/BMS/CMakeLists.txt
Comment (Jiajun-Huang) [reply firmware/thruna/BMS/CMakeLists.txt:65]: I think you are right. It should not be a problem to include more directories for `target_include_directories` than needed. I will make this as a variable in `firmware/Cmakelist.txt`, and remove this argument from the functions.

**Review by gtaharaedmonds — APPROVED** [review]
Nice!

**Pull request merged**
